### ISSUE TITLE
Training data format

### ISSUE THREAD

**Issue by TheGoblinTechies** [paul-krug/pytorch-tcn#16]
Hi Paul-krug,

Thanks for your nicely written TCN repo. I am quite new to this time-series prediction area, and I'm trying to use TCN as a robot manipulation controller, which would usually use this: robot_action_at_time_t = TCN(environment_observation_at_time_{t-1},..., at_time{t-m+1}) by using the historical m steps observation to predict the current action. I tried TCN with the input (1, 153, 1000), where 1 is the batch_size, 153 is the observation dimension, and 1000 is the total time step for a whole trajectory. The action dimension is 25, so the output_projection in TCN is 25. I noticed that the output is (1, 25, 1000), where 1 and 25 are understandable, and 1000 is the total time steps of the manipulation. I have the following questions:
1. During training, should I use a sub-trajectory with a moving window of fixed length (e.g. 50 steps for each, slicing the whole sequence 1000 into many small pieces) or simply use the whole sequence as input is recommended?
2. For those actions (predictions) in the early part of the trajectory (early part of the time steps, e.g. the second steps in 1000), are their input also got zero-padding (e.g. the second steps in 1000 are zero-padded with 998 zeros) as you described in the streaming inference? You mentioned: For kernel sizes > 1, a TCN will always use zero padding to ensure that the output has the same number of time steps as the input.
3. During inference (when the robot starts to manipulate), should I use the same number of time steps (sequence lengths, like 50 as a moving window input) as training and I manually do data shift to discard oldest observations and include newest observations? Will different numbers of sequence lengths influence the performance?
I really appreciate your replay! 
Zhaole

**Comment by paul-krug** [comment]
Sorry for the late reply, but I will try to answer your questions:

1. It is generally ok to use the whole sequence as input, but you may use shorter slices if your specific problem requires it.
2. Zero-padding is used in the convolutions in order to keep the amount of input and output time steps equal. No padding is required on your targets (unless you explicitly want to introduce some sort of past/future prediction).
3. If you trained a fully causal network, then it does not matter how many frames you give it at inference time, i.e. the prediction is the same regardless of whether it takes place on the entire sequence or frame by frame.

I hope that helps.
Cheers,
  Paul
